File: week01/red-sprout/solution03.py
maxSize = 0
cache = {}


def getOld():
    val = 1_000_000_007
    old = ""

    for key in cache:
        if val > cache[key]:
            val = cache[key]
            old = key

    return old


def update(city, idx):
    global maxSize, cache

    if maxSize == 0:
        return 5

    key = city.lower()
    t = 1 if key in cache else 5

    if key in cache or len(cache) < maxSize:
        cache[key] = idx
    else:
        del cache[getOld()]
        cache[key] = idx

    return t


def solution(cacheSize, cities):
    global maxSize, cache

    t = 0
    l = len(cities)

    maxSize = cacheSize
    cache = {}

    for i in range(l):
        t += update(cities[i], i)

    return t

File: week01/red-sprout/solution03.java
import java.util.*;

class Solution {
    int maxSize;
    Map<String, Integer> cache;

    String getOld() {
        int val = 1_000_000_007;
        String old = "";

        for(String key : cache.keySet()) {
            if(val > cache.get(key)) {
                val = cache.get(key);
                old = key;
            }
        }

        return old;
    }

    int update(String city, int idx) {
        if(maxSize == 0) return 5;
        String key = city.toLowerCase();
        int t = cache.containsKey(key) ? 1 : 5;

        if(cache.containsKey(key)
           || cache.size() < maxSize) {
            cache.put(key, idx);
        } else {
            cache.remove(getOld());
            cache.put(key, idx);
        }

        return t;
    }

    public int solution(int cacheSize, String[] cities) {
        int t = 0;
        int l = cities.length;
        maxSize = cacheSize;
        cache = new HashMap<>();

        for(int i = 0; i < l; ++i) {
            t += update(cities[i], i);
        }

        return t;
    }
}
